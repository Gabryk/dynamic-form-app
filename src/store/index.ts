import { configureStore } from "@reduxjs/toolkit";
import userApplicationReducer from "./userApplicationSlice";

export const reducer = { userApplication: userApplicationReducer };

export const store = configureStore({
  reducer,
});
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
