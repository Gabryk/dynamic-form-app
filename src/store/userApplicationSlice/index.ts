import { PayloadAction, createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import {
  RejectedActionFromAsyncThunk,
  FulfilledActionFromAsyncThunk,
  PendingActionFromAsyncThunk,
} from "@reduxjs/toolkit/dist/matchers";
import fakeApplication from "services/fakeApplication";

type Data = { [x: string]: string | number };
export interface UserApplication {
  loading?: boolean;
  error?: string;
  data?: Data;
}

const initialState: UserApplication = {};

export const uploadUserApplication = createAsyncThunk(
  "user/upload",
  async (data: Data, { signal }) => {
    const [promise, abort] = fakeApplication(data);

    signal.addEventListener("abort", () => {
      abort();
    });

    return await promise;
  }
);
const setSuccessUserApplication = (
  _: UserApplication,
  action: FulfilledActionFromAsyncThunk<typeof uploadUserApplication>
) => ({
  loading: false,
  data: action.payload,
});
const setLoadingUserApplication = (
  _: UserApplication,
  action: PendingActionFromAsyncThunk<typeof uploadUserApplication>
) => ({
  loading: true,
});
const setFailUserApplication = (
  _: UserApplication,
  action: RejectedActionFromAsyncThunk<typeof uploadUserApplication>
): UserApplication => ({
  loading: false,
  error: action.error.message,
});

const setUserData = (
  _: UserApplication,
  action: PayloadAction<Data>
): UserApplication => {
  return { data: action.payload, loading: false };
};

export const userApplicationSlice = createSlice({
  name: "user",
  initialState,
  reducers: {
    setUserData,
  },
  extraReducers(builder) {
    builder.addCase(uploadUserApplication.pending, setLoadingUserApplication);
    builder.addCase(uploadUserApplication.fulfilled, (state, action) =>
      setSuccessUserApplication(state, action)
    );
    builder.addCase(uploadUserApplication.rejected, setFailUserApplication);
  },
});

export const { setUserData: setUserDataAction } = userApplicationSlice.actions;

export default userApplicationSlice.reducer;
