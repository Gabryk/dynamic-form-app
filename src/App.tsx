import { useSelector } from "react-redux";
import Home from "pages/Home/Home";
import { RootState } from "store";
import Submited from "pages/Submited/Submited";
import ThemeProvider from "styles/ThemeProvider";

function App() {
  const { data } = useSelector((state: RootState) => state.userApplication);

  return (
    <ThemeProvider>
      {!Object.keys(data ?? {}).length ? <Home /> : <Submited />}
    </ThemeProvider>
  );
}

export default App;
