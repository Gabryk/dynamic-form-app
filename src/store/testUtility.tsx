import React, { PropsWithChildren } from "react";
import { combineReducers, configureStore } from "@reduxjs/toolkit";
import type { PreloadedState } from "@reduxjs/toolkit";
import type { RenderOptions } from "@testing-library/react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { reducer } from "store";

// Create the root reducer separately so we can extract the RootState type
const rootReducer = combineReducers(reducer);

const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

type RootState = ReturnType<typeof rootReducer>;
type AppStore = ReturnType<typeof setupStore>;
