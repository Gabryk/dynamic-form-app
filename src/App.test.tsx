import React from "react";
import { renderWithProviders } from "store/testUtility";
import App from "./App";

test("renders learn react link", () => {
  const { asFragment } = renderWithProviders(<App />);
  expect(asFragment()).toMatchSnapshot();
});
