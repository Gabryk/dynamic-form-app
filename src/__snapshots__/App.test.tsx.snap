// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-qRuGl dsiDPw"
    >
      <h1>
        Application Form
      </h1>
    </div>
    <form
      class="sc-iAEyCc WytVh"
    >
      <div
        class="sc-gLDxTj ejevhj"
      >
        <div
          class="sc-hLsdIM btFSFx"
        >
          <label
            class="sc-gueWre dVhXsQ"
            for="firstName"
          >
            First Name 
            <span
              class="sc-dmqGxT Ka-dMI"
            >
              *
            </span>
          </label>
          <input
            aria-describedby="exp-firstName"
            aria-required="true"
            class="sc-eDDNPB iOEymh"
            id="firstName"
            name="firstName"
            placeholder="First name"
            type="name"
          />
        </div>
        <div
          class="sc-hLsdIM btFSFx"
        >
          <label
            class="sc-gueWre dVhXsQ"
            for="lastName"
          >
            Last Name 
            <span
              class="sc-dmqGxT Ka-dMI"
            >
              *
            </span>
          </label>
          <input
            aria-describedby="exp-lastName"
            aria-required="true"
            class="sc-eDDNPB iOEymh"
            id="lastName"
            name="lastName"
            placeholder="Last name"
            type="name"
          />
        </div>
      </div>
      <div
        class="sc-hLsdIM btFSFx"
      >
        <label
          class="sc-gueWre dVhXsQ"
          for="email"
        >
          Email 
          <span
            class="sc-dmqGxT Ka-dMI"
          >
            *
          </span>
        </label>
        <input
          aria-describedby="exp-email"
          aria-required="true"
          class="sc-eDDNPB iOEymh"
          id="email"
          name="email"
          type="email"
        />
      </div>
      <div
        class="sc-hLsdIM btFSFx"
      >
        <label
          class="sc-gueWre dVhXsQ"
          for="address1"
        >
          Address 1 
        </label>
        <input
          aria-describedby="exp-address1"
          aria-required="false"
          class="sc-eDDNPB iOEymh"
          id="address1"
          name="address1"
          placeholder="Address 1"
          type="text"
        />
      </div>
      <div
        class="sc-gLDxTj ejevhj"
      >
        <div
          class="sc-hLsdIM btFSFx"
        >
          <label
            class="sc-gueWre dVhXsQ"
            for="city"
          >
            City 
          </label>
          <input
            aria-describedby="exp-city"
            aria-required="false"
            class="sc-eDDNPB iOEymh"
            id="city"
            name="city"
            type="text"
          />
        </div>
        <div
          class="sc-hLsdIM btFSFx"
        >
          <label
            class="sc-gueWre dVhXsQ"
            for="state"
          >
            State 
          </label>
          <input
            aria-describedby="exp-state"
            aria-required="false"
            class="sc-eDDNPB iOEymh"
            id="state"
            name="state"
            type="text"
          />
        </div>
        <div
          class="sc-hLsdIM btFSFx"
        >
          <label
            class="sc-gueWre dVhXsQ"
            for="zip"
          >
            ZIP 
          </label>
          <input
            aria-describedby="exp-zip"
            aria-required="false"
            class="sc-eDDNPB iOEymh"
            id="zip"
            name="zip"
            type="text"
          />
        </div>
      </div>
      <div
        class="sc-hLsdIM btFSFx"
      >
        <label
          class="sc-gueWre dVhXsQ"
          for="phone"
        >
          Phone number 
          <span
            class="sc-dmqGxT Ka-dMI"
          >
            *
          </span>
        </label>
        <input
          aria-describedby="exp-phone"
          aria-required="true"
          class="sc-eDDNPB iOEymh"
          id="phone"
          name="phone"
          placeholder="999 999 9999"
          type="phone"
        />
      </div>
      <div
        class="sc-hLsdIM btFSFx"
      >
        <label
          class="sc-gueWre dVhXsQ"
          for="jobTitle"
        >
          Job Title 
        </label>
        <select
          aria-required="false"
          class="sc-eDDNPB iOEymh"
          id="jobTitle"
          name="jobTitle"
          placeholder="Please select job title"
        >
          <option
            value="Engineer - lead"
          >
            Engineer - lead
          </option>
          <option
            value="Engineer - mid level"
          >
            Engineer - mid level
          </option>
          <option
            value="Engineer - junion"
          >
            Engineer - junion
          </option>
          <option
            value="Engineer - front end focused"
          >
            Engineer - front end focused
          </option>
          <option
            value="Engineer - backend focused"
          >
            Engineer - backend focused
          </option>
          <option
            value="Engineer - full stack"
          >
            Engineer - full stack
          </option>
        </select>
      </div>
      <div
        class="sc-hLsdIM btFSFx"
      >
        <label
          class="sc-gueWre dVhXsQ"
          for="reason"
        >
          Reason 
        </label>
        <textarea
          aria-required="false"
          class="sc-eDDNPB iOEymh"
          id="reason"
          name="reason"
          placeholder="Describe why you are a good fit for the job you are applying for."
        />
      </div>
      <button
        class="sc-jTrNMu kkVpuo"
      >
        Submit
      </button>
      <p>
        <span
          class="sc-dmqGxT Ka-dMI"
        >
          *
        </span>
        Required field
      </p>
    </form>
  </div>
</DocumentFragment>
`;
